autotest infrastructure improvements

peterbarker opened the pull request:
- makes changing mode much faster, which can help reduce impact of race conditions between suite and simulation
 - removes a pair of race conditions in some tests
 - make retrieving simulation time a little faster by accepting any of several messages
 - remove race condition around reboot where we can change mode so fast the simulation then considers the RC flight mode switch to be debounced and changes mode on us
 - remove some old and redundant mode-is functions

peterbarker committed:
autotest: make change_mode much faster

we don't need to know the current mode, so no need to wait for a heartbeat.

Send command after getting the current sim time so the insta-heartbeat-on-mode-change that ArduPilot sends out isn't swallowed by the get_sim_time call

peterbarker committed:
autotest: do not declare reboot complete until RC_CHANNELS received

There's a race condition around the autotest suite setting a mode on a newly-rebooted simulation vs the simulation deciding the flight-mode-channel is debounced.

peterbarker committed:
autotest: run mavlink rangefinder test at reduced speedup

the data can be timed out by the autopilot at higher speedups

peterbarker committed:
AP_HAL_SITL: check loop time more rapidly in threads

sleeping for 1ms limits the rate at which the logging thread (a real pthread in SITL!) can log.  If it only wakes up at 1kHz and only sends 1 4kB block each time it runs then that limits the logging rate to 4MB/second.  When we are running at over 40x speedup that means ArduPilot can only log at ~100kB/simulatedsecond - which is very close to our actual rate.  If an autotest tries to log faster (e.g test.Copter.FAST_ATTITUDE then we start dropping log messages.

peterbarker committed:
autotest: make DataFlashErase test more reliable

peterbarker commented:
Some of this is in #31899

khancyr commented on libraries/AP_HAL_SITL/SITL_State.cpp:
```diff
@@ -176,7 +176,7 @@ void SITL_State::wait_clock(uint64_t wait_time_usec)
             }
 #endif
-            usleep(1000);
+            usleep(10);
```
I am not sure most computer can catch up this, probably worth putting a comment to say that is a target but generally not achieved consistantly

peterbarker replied:
Done

khancyr reviewed (commented):
looks sane but break the tests

peterbarker committed:
auotest: do not lose heartbeats when arming

peterbarker committed:
AP_Logger: increase reserved space for non-messagewriter messages

when we go to write parameters out we can fill the buffer all the way up except for the 1kB we leave for things other than messagewriter messages.
That 1kB can then get filled quite quickly.

Push that little buffer out just a little bit to allow for fewer dropped messages at startup.

peterbarker committed:
Copter: correct initialisation issue in SITL

the truncation is not going to be a problem on a real vehicle, but the SITL-only sanity check where we go to use it in change_alt_frame is a problem.

khancyr commented:
@peterbarker could we extract AP_HAL_SITL: check loop time more rapidly in threads as a single PR to have integrated faster ? This is a good improvement on SITL timing and shouldn't be hold by other fix IMHO

peterbarker commented:
> @peterbarker could we extract AP_HAL_SITL: check loop time more rapidly in threads as a single PR to have integrated faster ? This is a good improvement on SITL timing and shouldn't be hold by other fix IMHO

That one actually thoroughly kills the DataFlash logging tests!  We start logging so much data that we swamp the test's acceptable boundaries.

As mentioned elsewhere, currently we drop something like 90% of the dataflash packets!

peterbarker commented:
This one fixes intermittent Replay failures.

peterbarker committed:
AP_HAL_SITL: comment the real-time-thread delay-for-time interval

IamPete1 commented on libraries/AP_Logger/AP_Logger_Backend.h:
Should we stick a intenal error in here? My read of the code is that someone would have to set there `LOG_FILE_BUFSIZE` to 4 or less for this to trigger. Looks like the Skyviper set it to 1?
